src/Stratis.Bitcoin.Features.Consensus/UnspentOutputSet.cs

```diff
@@ -60,7 +60,10 @@ public void SetCoins(UnspentOutputs[] coins)
             foreach (UnspentOutputs coin in coins)
             {
                 if (coin != null)
-                    this.unspents.Add(coin.TransactionId, coin);
+                {
+                    // We clone every UnspentOutputs as those will get mutated.
+                    this.unspents.Add(coin.TransactionId, coin.Clone());
+                }
```

Comment:
this is worse than our solution because now you don't have TrySetCoins covered below

Reply:
basically we were protecting every fetch, you are only protecting coin view rule. this means that someone working with fetched coin (e.g. miner) could have their coin modified silently without noticing

Reply:
yes correct you convinced me I will change back